src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -74,31 +412,41 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
   // Free set maintains mutator and collector views, and normally they allocate in their views only,
   // unless we special cases for stealing and mixed allocations.
 
+  // Overwrite with non-zero (non-NULL) values only if necessary for allocation bookkeeping.
+
   switch (req.type()) {
     case ShenandoahAllocRequest::_alloc_tlab:
     case ShenandoahAllocRequest::_alloc_shared: {
-
       // Try to allocate in the mutator view
-      for (size_t idx = _mutator_leftmost; idx <= _mutator_rightmost; idx++) {
-        if (is_mutator_free(idx)) {
-          HeapWord* result = try_allocate_in(_heap->get_region(idx), req, in_new_region);
-          if (result != nullptr) {
-            return result;
+      // Allocate within mutator free from high memory to low so as to preserve low memory for humongous allocations
+      if (!_free_sets.is_empty(Mutator)) {
+        // Use signed idx.  Otherwise, loop will never terminate.
+        int leftmost = (int) _free_sets.leftmost(Mutator);
+        for (int idx = (int) _free_sets.rightmost(Mutator); idx >= leftmost; idx--) {
+          ShenandoahHeapRegion* r = _heap->get_region(idx);
+          if (_free_sets.in_free_set(idx, Mutator)) {
+            // try_allocate_in() increases used if the allocation is successful.
+            HeapWord* result;
+            size_t min_size = (req.type() == ShenandoahAllocRequest::_alloc_tlab)? req.min_size(): req.size();
+            if ((alloc_capacity(r) >= min_size) && ((result = try_allocate_in(r, req, in_new_region)) != nullptr)) {
+              return result;
+            }
           }
         }
       }
-
       // There is no recovery. Mutator does not touch collector view at all.
       break;
     }
     case ShenandoahAllocRequest::_alloc_gclab:
+      // GCLABs are for evacuation so we must be in evacuation phase.  If this allocation is successful, increment
+      // the relevant evac_expended rather than used value.
+
```

Comment:
This seems like a leakage of abstraction to me. Besides, I do not see `evac_expended` anywhere in this file?

Reply:
Removing mention of evac_expended.  That comment is vestigial.